wagtail/wagtailcore/blocks/collector.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import, unicode_literals
+
+import re
+from warnings import warn
+
+from django.apps import apps
+from django.db.models import Model, Q
+
+from wagtail.wagtailcore.blocks import ListBlock, StreamBlock, StreamValue, StructBlock, StructValue
+from wagtail.wagtailcore.fields import StreamField
+
+
+class StreamFieldCollector:
+    def __init__(self, field):
+        self.field = field
+
+    def find_block_type(self, block_type, blocks=None, block_path=None):
+        if blocks is None:
+            blocks = [self.field.stream_block]
+        if block_path is None:
+            block_path = [self.field.stream_block]
+        for block in blocks:
+            current_path = block_path + [block]
+            if isinstance(block, block_type):
+                return current_path
+            elif isinstance(block, (StreamBlock, StructBlock)):
+                return self.find_block_type(
+                    block_type, block.child_blocks.values(),
+                    block_path=current_path)
+            elif isinstance(block, ListBlock):
+                return self.find_block_type(block_type, [block.child_block],
+                                            block_path=current_path)
+
+    def find_values(self, value, block_path):
+        block_path = block_path[1:]
+        if len(block_path) == 1:
+            last_block = block_path[0]
+            if isinstance(value, StreamValue.StreamChild):
+                if last_block.name == value.block.name:
+                    yield value.value
+            elif isinstance(value, StructValue):
+                if last_block.name in value:
+                    yield value[last_block.name]
+        else:
+            if isinstance(value, StreamValue):
+                for item in value:
+                    yield from self.find_values(item, block_path)
+            elif isinstance(value, StreamValue.StreamChild):
+                if value.block.name == value.block.name:
+                    yield from self.find_values(value.value, block_path)
+            else:
+                warn('Unexpected StreamField value: <%s: %s>'
+                     % (type(value), str(value)[:30]))
+
+
+class ModelStreamFieldsCollector:
+    def __init__(self, model):
+        self.model = model
+        self.fields = [field for field in model._meta.fields
+                       if isinstance(field, StreamField)]
+        self.field_collectors = [StreamFieldCollector(field)
+                                 for field in self.fields]
+
+    def prepare_value(self, value):
+        if isinstance(value, Model):
+            return str(value.pk)
+        if isinstance(value, (int, float)):
+            return str(value)
+        if isinstance(value, bool):
+            return 'true' if value else 'false'
+        if isinstance(value, None):
+            return 'null'
+        return '"%s"' % re.escape(str(value).replace('"', r'\"')
+                                  .replace('|', r'\|'))
```

Comment:
Almost all this could probably be replaced with `json.dumps(…)`. I wrote it this way since it seemed more flexible if I needed some special case like models.